### DISCUSSION TITLE
Set different units for length, altitude, speed

### DISCUSSION THREAD

**Discussion by MotoUKRider** [devemux86/cruiser#200] [Q&A] [answered]
On a recent ride using GraphHopper (I'm really a BRouter man, honest) - I enabled the "Speed" display on my navigation pane.  The road limit was not present on some roads, but it was on others.  CoPilot has a similar feature (a red speed limit sign overlay) which works on all roads I've travelled on.  That is a minor issue, it's just nice to have it.

Oddly, using the UK's imperial "MPH" unit - I frequently saw "59" instead of "60".  Something to be aware of as I *think* these values are converted from metric.  Not knowing anything about how GrapHopper is built, but it'd be interesting to inspect a routing file and see the information :)

Note: We are odd here in the UK, we mix our units!

**Comment by devemux86** [comment]
@MotoUKRider thanks for the report!

[GraphHopper documentation](https://docs.graphhopper.com/) mentions: _"max_speed: the speed limit from a sign (km/h)"_

Cruiser internally makes the conversion to mph.
Maybe we need to round the result, 59 mph -> 60 mph?

**Reply by devemux86** [reply]
Can you post the OpenStreetMap link of such road to see what is declared in the OpenStreetMap database?

**Reply by MotoUKRider** [reply]
Yes, I was on this road - and digging into the OSM I can see that some parts of the road have a speed "tag", others do not.
https://www.openstreetmap.org/way/227103368 and onto https://www.openstreetmap.org/way/62066115#map=14/53.3978/-0.0112

In simulation via Cruiser app:
Here, I saw 31mph: https://www.openstreetmap.org/way/115469057
Further down the road, in Utterby, I saw 50mph and then 40mph where the limits change (all OK).
Speed limit vanished here: https://www.openstreetmap.org/way/62066115#map=17/53.40316/-0.02673

Certainly the numeric values need be "handled" *if* being converted.

**Comment by devemux86** [comment]
Next Cruiser with imperial units will round the speed limits.
Note: speed limits are provided in metric units by the routers.

@MotoUKRider this will solve such cases?

Your example has 60 mph in OpenStreetMap database.
It is probably converted to 95 km/h in GraphHopper / Kurviger.
Cruiser gets 95 km/h (59 mph) from GraphHopper / Kurviger.
Now it will return the rounded value of 60 mph.

**Reply by MotoUKRider** [reply]
Thanks for this - Now that I can see how to query the map by clicking on the road-name in OSM, I can see lots of places where no speed limit is offered.  This may explain why the value sometimes disappears from Cruiser during navigation?  eg: It completely misses a 30-60mph residential-to-rural change in a few places, and a 2yr-old revised speed-limit (40 to 30) so the source data isn't perfect and I'm not sure if I fancy getting involved with editing/fixing them!

This only really came to light because I wanted to test GraphHopper routing with the latest build (gates etc) and knew that "Speed" was available for display - it was an interesting experience.

Looking at road "types" vs speeds - I can see from some local examples that assuming a speed based on "residential", "secondary" etc is probably not a good idea.  One such example road has a long stretch of 60, then 30, then a short 40 and back up to 60 before dropping to 40.  The data is probably about as good as it'll get unless keen people get stuck in.

I can see why a couple of users run that overlay app with the "speed sign" :)

**Reply by devemux86** [reply]
> This may explain why the value sometimes disappears from Cruiser during navigation?

Speed limits are provided in the following order:
OpenStreetMap database -> Routing service -> Applications

If one of the first two steps does not provide the information,
apps cannot receive and display speed limits.

**Reply by devemux86** [reply]
> I can see from some local examples that assuming a speed based on "residential", "secondary" etc is probably not a good idea.

Indeed when there is no max speed declared in OpenStreetMap,
some routers and apps derive the max speed from the road type.

But it is not very safe, as countries have different speed limits:
https://en.wikipedia.org/wiki/Speed_limits_by_country

**Comment by devemux86** [comment]
> Next Cruiser with imperial units will round the speed limits.

Available in [Cruiser 3.0.23](https://play.google.com/store/apps/details?id=gr.talent.cruiser).

**Comment by MotoUKRider** [comment]
Coming back to this, during my recent "down the rabbit-hole" of nav apps on the Play Store, one particular app (I forget which one) had a "Unit" setting for each value (rather than a simple Imperial/Metric switch) where awkward people :raised_hand_with_fingers_splayed: could use MPH for speed and metres for GPS height, plus the formatting of Lat/Lon and any other measurement format/unit.

**Reply by devemux86** [reply]
Some apps allow different units for length and speed.
Do we need such separate options?

Coordinate format options are already available in Cruiser.

**Reply by MotoUKRider** [reply]
@devemux86 For me, Speed/Height are the important ones.  There's a mix of units here in the UK, and it also depends on your age!

**Reply by devemux86** [reply]
@MotoUKRider What about the imperial units?

OsmAnd has some mixed combinations: Kilometers/meters, Miles/feet, Miles/meters, Miles/yards.

For speed km/h and mph are ok?

Also horizontal distance can be measured in kilometers / miles.
While vertical distance (height) is usually measured in meters / feet.
For example [Everest](https://en.wikipedia.org/wiki/Mount_Everest) 8,848.86 m (29,031.7 ft)

**Reply by devemux86** [reply]
Note that all systems (online and offline) work with the metric system, i.e. meters and m/s.
We do all the conversions on top, that's why distance and speed sliders display both units.

**Reply by MotoUKRider** [reply]
@devemux86 It's not a set of choices you'd change much, but having the ability to define what unit of measurement you see (and hear) could be beneficial.  The OSMand settings work for me, but I can't remember what the app was that had individual options rather than a "Regional" series of presets.

I'm not saying that we need *"In 0.0000001 light-years, turn right"* announcements, just a consideration of adjusting units.

**Reply by devemux86** [reply]
OsmAnd mixes the metric and imperial units. Is this mix used?

Or we can just split the units setting to distance and speed units?
- Distance: km/m, mi/ft, nmi/m
- Speed: km/h, mph, kn

Are the yards used?

**Reply by MotoUKRider** [reply]
Miles/metres is what I have OSMand set to - I don't use the app much but it's on the phone anyway.
You also have the situation where driving/riding in the UK (or Europe) that you may need to switch the speed unit but prefer to stick with GPS Height in the unit you're familiar with.

Yards are used in the UK to count down to a motorway junction, /// , // , / so there's a bit of everything.

**Comment by devemux86** [comment]
The nautical units exist because some time ago I developed "Nautilus Charts".
An offline nautical charts plotting and positioning system, based on Cruiser.

![Nautilus Charts](https://github.com/devemux86/cruiser/assets/3484020/1aa1e70d-9045-413e-a179-85bbf444b4a0)

**Comment by devemux86** [comment, marked as answer]
- Implemented in https://github.com/devemux86/cruiser/discussions/891
